source/mir/math/stat.d

```diff
@@ -1626,6 +1627,704 @@ unittest
+    })) {
+        return x * x;
+    } else {
+        static assert(0, "square: cannot square type: " ~ T.stringof);
+    }
+}
+
+@safe pure nothrow @nogc
+unittest {
+    int x = square(4);
+    assert(x == 16);
+    double y = square(5.0);
+	assert(y == 25);
+    
+    static assert(is(typeof(square(4)) == int));
+    static assert(is(typeof(square(5.0)) == double));
+}
+
+private
+mixin template variance_ops(T, Summation summation)
+{
+    import core.lifetime: move;
+
+    ///
+    MeanAccumulator!(T, summation) meanAccumulator;
+
+    ///
+    size_t count() @property
+    {
+        return meanAccumulator.count;
+    }
+}
+
+///
+struct VarianceAccumulator(T, VarianceAlgo varianceAlgo, Summation summation)
+    if (isMutable!T && varianceAlgo == VarianceAlgo.naive)
+{
+    mixin variance_ops!(T, summation);
+
+    ///
+    MapSummator!(square, T, summation) sumOfSquares;
+    
+    ///
+    void put(Range)(Range r)
+        if (isIterable!Range)
+    {
+        static if (hasShape!Range)
+        {
+            meanAccumulator.put(r.move);
+            sumOfSquares.put(r.move);
+        }
+        else
+        {
+            foreach(x; r)
+            {
+                meanAccumulator.put(x);
+                sumOfSquares.put(x);
+            }
+        }
+    }
+
+    ///
+    void put()(T x)
+    {
+        meanAccumulator.put(x);
+        sumOfSquares.put(x);
+    }
+    
+    ///
+    F variance(F = T)(bool isPopulation) @property
+    {
+        if (isPopulation == false)
+            return cast(F) sumOfSquares.sum / cast(F) (count - 1) - 
+                square(cast(F) meanAccumulator.mean) * (cast(F) count / cast(F) (count - 1));
+        else
+            return cast(F) sumOfSquares.sum / cast(F) count - 
+                square(cast(F) meanAccumulator.mean);
+    }
+}
+
+///
+version(mir_test)
+@safe pure nothrow
+unittest
+{
+    import mir.ndslice.slice: sliced;
+    import mir.math.common: approxEqual;
+
+    auto x = [0.0, 1.0, 1.5, 2.0, 3.5, 4.25,
+              2.0, 7.5, 5.0, 1.0, 1.5, 0.0].sliced;
+
+    enum PopulationTrueCT = true;
+    enum PopulationFalseCT = false;
+    bool PopulationTrueRT = true;
+    bool PopulationFalseRT = false;
+
+    VarianceAccumulator!(double, VarianceAlgo.naive, Summation.naive) v;
+    v.put(x);
+    assert(v.variance(PopulationTrueRT).approxEqual(54.76562 / 12));
+    assert(v.variance(PopulationTrueCT).approxEqual(54.76562 / 12));
+    assert(v.variance(PopulationFalseRT).approxEqual(54.76562 / 11));
+    assert(v.variance(PopulationFalseCT).approxEqual(54.76562 / 11));
+
+    v.put(4.0);
+    assert(v.variance(PopulationTrueRT).approxEqual(57.01923 / 13));
+    assert(v.variance(PopulationTrueCT).approxEqual(57.01923 / 13));
+    assert(v.variance(PopulationFalseRT).approxEqual(57.01923 / 12));
+    assert(v.variance(PopulationFalseCT).approxEqual(57.01923 / 12));
+}
+
+///
+struct VarianceAccumulator(T, VarianceAlgo varianceAlgo, Summation summation)
```

Comment:
The python script in the wikipedia is more performance. It is better to use raw summators for `mean` and `M2` instead of some special accumulators. This will allow do not do branching for `count == 0`